source/common/router/router.cc

```diff
@@ -328,6 +336,29 @@ void Filter::chargeUpstreamCode(Http::Code code,
   chargeUpstreamCode(response_status_code, *fake_response_headers, upstream_host, dropped);
 }
 
+void Filter::finalizeTimeoutHeaders() {
+  // if we have a deadline, update the timeout headers.
+  if (deadline_) {
+    // calculate how much time we have left
+    std::chrono::milliseconds expected_timeout =
+        std::chrono::duration_cast<std::chrono::milliseconds>(deadline_.value() -
+                                                              config_.timeSource().monotonicTime());
+    // this _should_ be positive
+    // TODO: what should be done if it is not positive? as i believe this can happen in rare
+    // cases.
+    ASSERT(expected_timeout.count() > 0, "expected_timeout is not positive");
+    // if it is less than the current per try timeout, or there is no per try timeout, update
+    // timeout headers
+    if (timeout_.per_try_timeout_.count() == 0 || expected_timeout < timeout_.per_try_timeout_) {
+      FilterUtility::setTimeoutHeaders(*route_entry_, expected_timeout.count(), *downstream_headers_,
```

Comment:
Even in cases where we start measuring in request complete, shouldn't we still adjust expected timeout HTTP and gRPC time based on how long it took to get a connection from the CP? I think this has been raised elsewhere and it would be good to fix this also?

Reply:
i can adjust; my intention was to prevent a change in behavior to reduce risk

Reply:
I think it would be fine to do this as a separate change, just something to keep in mind.